src/main/kotlin/parts/wisdom/arcraider/visualization/TaskChooser.kt

```diff
@@ -0,0 +1,87 @@
+package parts.wisdom.arcraider.visualization
+
+import java.awt.Dimension
+import java.io.File
+import javax.swing.JPanel
+import javax.swing.JScrollPane
+import javax.swing.JTree
+import javax.swing.event.TreeExpansionEvent
+import javax.swing.event.TreeExpansionListener
+import javax.swing.tree.DefaultMutableTreeNode
+import javax.swing.tree.DefaultTreeModel
+
+class FileBrowser(dir: String, onFileSelected: (File) -> Unit = { }) : JPanel() {
+
+    private lateinit var tree: JTree
+
+    private val nodeExpansionListener = object: TreeExpansionListener {
+        override fun treeExpanded(event: TreeExpansionEvent?) {
+            val subtree = (event?.path?.lastPathComponent as? DefaultMutableTreeNode) ?: return
+            if (subtree.depth < 2) {
+                for (child in subtree.children()) {
+                    TreeExplorer(child as DefaultMutableTreeNode).exploreNodes(3)
+                }
+            }
+        }
+
+        override fun treeCollapsed(event: TreeExpansionEvent?) = Unit
+    }
+
+    init {
+        val rootFile = File(dir)
+        val rootNode = DefaultMutableTreeNode(RootFileNode(rootFile))
+        tree = JTree(DefaultTreeModel(rootNode))
+        tree.showsRootHandles = true
+        tree.addTreeExpansionListener(nodeExpansionListener)
+        tree.addTreeSelectionListener {
+            val node = tree.lastSelectedPathComponent as? DefaultMutableTreeNode
+            if (node?.isLeaf == true) {
+                onFileSelected(node.file())
+            }
+        }
+
+        val scrollPane = JScrollPane(tree).apply {
+            preferredSize = Dimension(200, 500)
+        }
+        add(scrollPane)
+        TreeExplorer(rootNode).exploreNodes(3)
+    }
+
+    private inner class TreeExplorer(
+            private val root: DefaultMutableTreeNode
+    ) {
+
+        fun exploreNodes(depth: Int) {
+            createChildren(root, depth)
+        }
+
+        private fun createChildren(node: DefaultMutableTreeNode, depth: Int) {
+            if (depth <= 0) return
+            // List all json files in directory, excluding hidden files.
+            val files = node.file().listFiles { file ->
+                !file.name.startsWith(".") && (file.isDirectory || file.extension == "json")
+            } ?: return
+            // Alphabetize the files to make it easier to look through.
+            files.sort()
+            for (file in files) {
+                val childNode = DefaultMutableTreeNode(FileNode(file))
+                node.add(childNode)
```

Comment:
Why doesn't this create duplicate child nodes? (I see from trying out the UI that it doesn't.)

Reply:
We list the files in a directory and then add each of them as a node in the tree. If a file is a directory, then we recurse with that as our "root" node and list its files/directories. Up to a depth of 3 (for perf reasons, we don't want to enumerate the entire disk)

Reply:
But then we are doing that again from nodeExpansionListener. I'm assuming that's invoked every time I close and re-open a tree node? But then I don't see why it doesn't create duplicate nodes?